lib/elixir/lib/macro.ex

```diff
@@ -298,6 +298,86 @@ defmodule Macro do
   def pipe_warning(_), do: nil
 
   @doc """
+  Rewrites an expression so it can be used both inside and outside a guard.
+
+  Take, for example, the expression:
+
+  ```elixir
+    is_integer(value) and rem(value, 2) == 0
+  ```
+
+  If we wanted to create a macro, `is_even`, from this expression, that could be
+  used in guards, we'd have to take several things into account.
+
+  First, if this expression is being used inside a guard, `value` needs to be
+  unquoted each place it occurs, since it has not yet been at that point in our
+  macro.
+
+  Secondly, if the expression is being used outside of a guard, we want to unquote
+  `value`––but only once, and then re-use the unquoted form throughout the expression.
+
+  This helper does exactly that: takes the AST for an expression and a list of
+  variable names it should be aware of, and rewrites it into a new expression that
+  checks for its presence in a guard, then unquotes the variable references as
+  appropriate.
+
+  The resulting transformation looks something like this:
+
+      > expr = quote do: is_integer(value) and rem(value, 2) == 0
+      > vars = [quote(do: value)]
+      > expr |> Macro.guard(vars) |> Macro.to_string |> IO.puts
+
+      case Macro.Env.in_guard? __CALLER__ do
+        true -> quote do
+          is_integer(unquote(value)) and rem(unquote(value), 2) == 0
+        end
+        false -> quote do
+          value = unquote(value)
+          is_integer(value) and rem(value, 2) == 0
+        end
+      end
+  """
+  @spec guard(Macro.t, [Macro.t]) :: Macro.t | no_return
```

Comment:
`Kernel.Utils` may be a better place for putting this, since I don't believe this should be public.